server/src/main/java/org/opensearch/action/search/TransportDeletePITAction.java

```diff
@@ -0,0 +1,172 @@
+                    );
+                    listener.onResponse(new DeletePITResponse(false));
+                }
+            }, e -> {
+                logger.debug("Delete PITs failed ", e);
+                listener.onResponse(new DeletePITResponse(false));
+            }));
+        }
+    }
+
+    /**
+     * Delete all active PIT reader contexts
+     */
+    void deleteAllPits(ActionListener<DeletePITResponse> listener) {
+        int size = clusterService.state().getNodes().getSize();
+        ActionListener groupedActionListener = getGroupedListener(listener, size);
+        for (final DiscoveryNode node : clusterService.state().getNodes()) {
+            try {
+                Transport.Connection connection = searchTransportService.getConnection(null, node);
+                searchTransportService.sendDeleteAllPitContexts(connection, groupedActionListener);
+            } catch (Exception e) {
+                groupedActionListener.onFailure(e);
+            }
+        }
+    }
+
+    /**
+     * Delete list of pits, return success if all reader contexts are deleted ( or not found ).
+     */
+    void deletePits(List<SearchContextIdForNode> contexts, ActionListener<Integer> listener) {
+        final StepListener<BiFunction<String, String, DiscoveryNode>> lookupListener = getLookupListener(contexts);
+        lookupListener.whenComplete(nodeLookup -> {
+            final GroupedActionListener<Boolean> groupedListener = new GroupedActionListener<>(
+                ActionListener.delegateFailure(
+                    listener,
+                    (l, result) -> l.onResponse(Math.toIntExact(result.stream().filter(r -> r).count()))
+                ),
+                contexts.size()
+            );
+
+            for (SearchContextIdForNode contextId : contexts) {
+                final DiscoveryNode node = nodeLookup.apply(contextId.getClusterAlias(), contextId.getNode());
+                if (node == null) {
+                    groupedListener.onFailure(new OpenSearchException("node not connected"));
+                } else {
+                    try {
+                        final Transport.Connection connection = searchTransportService.getConnection(contextId.getClusterAlias(), node);
+                        searchTransportService.sendPitFreeContext(
+                            connection,
+                            contextId.getSearchContextId(),
+                            ActionListener.wrap(r -> groupedListener.onResponse(r.isFreed()), e -> groupedListener.onResponse(false))
+                        );
+                    } catch (Exception e) {
+                        logger.debug("Delete PIT failed ", e);
+                        groupedListener.onResponse(false);
+                    }
+                }
+            }
+        }, listener::onFailure);
+    }
+
+    private StepListener<BiFunction<String, String, DiscoveryNode>> getLookupListener(List<SearchContextIdForNode> contexts) {
```

Comment:
this seems like a reusable method between create/delete/pit or on those lines. redundant? Maybe all in one service class can harness common functionalities better?

Reply:
not reusable since signature and hence logic is different between create and delete.